paraglob.cpp

```diff
@@ -0,0 +1,111 @@
+#include "paraglob.h"
+
+Paraglob::Paraglob(const std::vector<std::string> &patterns): Paraglob() {
+  for (std::string pattern : patterns) {
+    this->add(pattern);
+  }
+  this->compile();
+}
+
+void Paraglob::add(std::string pattern) {
+  // Get the meta words
+  std::vector<std::string> m_words = get_meta_words(pattern);
+  // Put them away
+  if (m_words.size() == 0) {
+    this->single_wildcards.push_back(pattern);
+  }
+  for (std::string meta_word : m_words) {
+    AhoCorasickPlus::EnumReturnStatus status;
+    AhoCorasickPlus::PatternId patId = this->meta_words.size();
+
+    status = this->my_ac.addPattern(meta_word, patId);
+    if (status != AhoCorasickPlus::RETURNSTATUS_SUCCESS &&
+      status != AhoCorasickPlus::RETURNSTATUS_DUPLICATE_PATTERN) {
+      std::cout << "Failed to add: " << meta_word << std::endl;
+    } else {
+      if (status == AhoCorasickPlus::RETURNSTATUS_DUPLICATE_PATTERN) {
+        this->meta_to_node_map.at(meta_word).add_pattern(pattern);
+      } else {
+        this->meta_words.push_back(meta_word);
+        this->meta_to_node_map.emplace(meta_word, ParaglobNode(meta_word, pattern));
+      }
+    }
+  }
+}
+
+void Paraglob::compile() {
+  this->my_ac.finalize();
+}
+
+std::vector<std::string> Paraglob::get(std::string text) {
+  // Narrow to the meta-word matches
+  AhoCorasickPlus::Match aMatch;
+  this->my_ac.search(text, false);
+  std::vector<std::string> meta_matches;
+  while (this->my_ac.findNext(aMatch)) {
+    meta_matches.push_back(this->meta_words.at(aMatch.id));
+  }
+
+  // Get the relevant patterns
+  std::vector<std::string> patterns;
+  for (std::string meta_word : meta_matches) {
+    this->meta_to_node_map.at(meta_word).merge_into(patterns);
```

Comment:
Could we run fnmatch() on the elements right here and avoid having to remove them later?

Reply:
Yes! That's clever. Doing this sped it up quite a bit.